### PR TITLE
feat: use paramiko for the ssh interactive session

### PR DESCRIPTION
**Pull request by ptrcnull**
i have **not** tested this code. please do.

(but for all i know, it should work; after all, it works on my machine outside vixos codebase)

### CODE REVIEW

**Comment by ptrcnull** [comment]
if this is meant to be used in a process with a longer lifetime, it probably needs unbinding of the signal handler

**Comment by ptrcnull** [comment]
added it just in case:

```python
signal.signal(signal.SIGWINCH, signal.SIG_DFL)
```
https://github.com/Patient-Engineering/vixos/pull/44/files#diff-b29bf4c33b7ef6732d74f97f7704eeb19c9e45a96dd0e7d0137db7b8a1a8b5e3R68

also, added some small comments

File: vixos/ssh.py [hunk 1]
```diff
@@ -23,20 +30,44 @@ def pubkey_text(self) -> str:
...
-                "StrictHostKeyChecking=no",
-                "-o",
-                "UserKnownHostsFile=/dev/null",
-            ]
-        )
+        ssh = self.ssh_session(user, host)
+        width, height = os.get_terminal_size()
+        shell_chan = ssh.invoke_shell(os.getenv('TERM', 'vt100'), width, height)
+
+        def signal_handler():
```
Comment (msm-code) [inline vixos/ssh.py]:
```suggestion
        def signal_handler(_signum, _frame):
```

File: vixos/ssh.py [hunk 2]
```diff
@@ -23,20 +30,44 @@ def pubkey_text(self) -> str:
...
+
+            while True:
+                read_ready, _, _ = select.select([shell_chan, sys.stdin], [], [])
+                if shell_chan in read_ready:
+                    try:
+                        b = shell_chan.recv(1024)
+                        if len(b) == 0:
+                            print(f'\nConnection to {host} closed')
+                            break
+                        sys.stdout.write(b)
```
Comment (msm-code) [inline vixos/ssh.py]:
```suggestion
                        sys.stdout.buffer.write(b)
```
(to avoid encoding errors)

**Review by msm-code — CHANGES_REQUESTED** [review]
Thanks! Some small comments

File: vixos/ssh.py [hunk 3]
```diff
@@ -23,20 +30,44 @@ def pubkey_text(self) -> str:
...
+                        sys.stdout.flush()
+                    except socket.timeout:
+                        pass
+                if sys.stdin in read_ready:
+                    b = sys.stdin.read(1)
+                    if len(b) == 0:
+                        break
+                    shell_chan.send(b)
```
Comment (msm-code) [inline vixos/ssh.py:65]: This is also a bytes/str mismatch. I guess it makes sense to read bytes from the stdin? If so, probably worth changing read to `sys.stdin.buffer.read()` to read raw bytes from stdin.

Comment (ptrcnull) [reply vixos/ssh.py:65]: hmm, yes, that is indeed another bytes/str mismatch; pushed an amended commit to fix all current issues

**Review by msm-code — APPROVED** [review]
:wizard:

**Pull request merged**